data/transactions/logic/eval.go

```diff
@@ -0,0 +1,1111 @@
+}
+
+func opLe(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	cond := cx.stack[prev].Uint <= cx.stack[last].Uint
+	if cond {
+		cx.stack[prev].Uint = 1
+	} else {
+		cx.stack[prev].Uint = 0
+	}
+	cx.stack = cx.stack[:last]
+}
+
+func opGe(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	cond := cx.stack[prev].Uint >= cx.stack[last].Uint
+	if cond {
+		cx.stack[prev].Uint = 1
+	} else {
+		cx.stack[prev].Uint = 0
+	}
+	cx.stack = cx.stack[:last]
+}
+
+func opAnd(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	cond := (cx.stack[prev].Uint != 0) && (cx.stack[last].Uint != 0)
+	if cond {
+		cx.stack[prev].Uint = 1
+	} else {
+		cx.stack[prev].Uint = 0
+	}
+	cx.stack = cx.stack[:last]
+}
+
+func opOr(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	cond := (cx.stack[prev].Uint != 0) || (cx.stack[last].Uint != 0)
+	if cond {
+		cx.stack[prev].Uint = 1
+	} else {
+		cx.stack[prev].Uint = 0
+	}
+	cx.stack = cx.stack[:last]
+}
+
+func opEq(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	ta := cx.stack[prev].argType()
+	tb := cx.stack[last].argType()
+	if ta != tb {
+		cx.err = fmt.Errorf("cannot compare (%s == %s)", cx.stack[prev].typeName(), cx.stack[last].typeName())
+		return
+	}
+	var cond bool
+	if ta == StackBytes {
+		cond = bytes.Compare(cx.stack[prev].Bytes, cx.stack[last].Bytes) == 0
+	} else {
+		cond = cx.stack[prev].Uint == cx.stack[last].Uint
+	}
+	if cond {
+		cx.stack[prev].Uint = 1
+	} else {
+		cx.stack[prev].Uint = 0
+	}
+	cx.stack[prev].Bytes = nil
+	cx.stack = cx.stack[:last]
+}
+
+func opNeq(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	ta := cx.stack[prev].argType()
+	tb := cx.stack[last].argType()
+	if ta != tb {
+		cx.err = fmt.Errorf("cannot compare (%s == %s)", cx.stack[prev].typeName(), cx.stack[last].typeName())
+		return
+	}
+	var cond bool
+	if ta == StackBytes {
+		cond = bytes.Compare(cx.stack[prev].Bytes, cx.stack[last].Bytes) != 0
+		cx.stack[prev].Bytes = nil
+	} else {
+		cond = cx.stack[prev].Uint != cx.stack[last].Uint
+	}
+	if cond {
+		cx.stack[prev].Uint = 1
+	} else {
+		cx.stack[prev].Uint = 0
+	}
+	cx.stack = cx.stack[:last]
+}
+
+func opNot(cx *evalContext) {
+	last := len(cx.stack) - 1
+	cond := cx.stack[last].Uint == 0
+	if cond {
+		cx.stack[last].Uint = 1
+	} else {
+		cx.stack[last].Uint = 0
+	}
+}
+
+func opLen(cx *evalContext) {
+	last := len(cx.stack) - 1
+	cx.stack[last].Uint = uint64(len(cx.stack[last].Bytes))
+	cx.stack[last].Bytes = nil
+}
+
+func opItob(cx *evalContext) {
+	last := len(cx.stack) - 1
+	ibytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(ibytes, cx.stack[last].Uint)
+	cx.stack[last].Bytes = ibytes
+}
+
+func opBtoi(cx *evalContext) {
+	last := len(cx.stack) - 1
+	ibytes := cx.stack[last].Bytes
+	if len(ibytes) > 8 {
+		cx.err = fmt.Errorf("btoi arg too long, got [%d]bytes", len(ibytes))
+		return
+	}
+	value := uint64(0)
+	for _, b := range ibytes {
```

Comment:
Is there a reason that you didn't do a symmetric `binary.BigEndian.Uint64` here? Also, what's the rationale behind using big-endian encoding?

Reply:
One thing that is possible but perhaps undesirable to support if we have little-endian encoding: `btoi` can accept slices longer than 8 bytes as long as bytes 9-end are all 0. This may be undesirable because `btoi(itob(data)) != data`.

Reply:
I still think of big endian as 'network order' (Thanks, SUN!)
I think `binary.BigEndian.Uint64()` chokes if the source isn't a full 8 bytes, but I wanted to be more flexible.